File: benchmarks-QuTiP/timeevolution_master_particle.py
import qutip as qt
import numpy as np
import benchmarkutils

name = "timeevolution_master_particle"

samples = 3
evals = 1
cutoffs = range(10, 41, 10)


def setup(N):
    xmin = -5
    xmax = 5
    x0 = 0.3
    p0 = -0.2
    sigma0 = 1
    dx = (xmax - xmin) / N
    pmin = -np.pi / dx
    pmax = np.pi / dx
    dp = (pmax - pmin) / N

    samplepoints_x = np.linspace(xmin, xmax, N, endpoint=False)
    samplepoints_p = np.linspace(pmin, pmax, N, endpoint=False)

    x = qt.Qobj(np.diag(samplepoints_x))
    row0 = [sum([p*np.exp(-1j*p*dxji) for p in samplepoints_p])*dp*dx/(2*np.pi) for dxji in samplepoints_x - xmin]
    row0 = np.array(row0)
    col0 = row0.conj()

    a = np.zeros([N, N], dtype=complex)
    for i in range(N):
        a[i, i:] = row0[:N - i]
        a[i:, i] = col0[:N - i]
    p = qt.Qobj(a)

    H = p**2 + 2 * x**2

    def gaussianstate(x0, p0, sigma0):
        alpha = 1./(np.pi**(1/4)*np.sqrt(sigma0))*np.sqrt(dx)
        data = alpha*np.exp(1j*p0*(samplepoints_x-x0/2) - (samplepoints_x-x0)**2/(2*sigma0**2))
        return qt.Qobj(data)

    psi0 = gaussianstate(x0, p0, sigma0)
    J = [x + 1j * p]

    options = qt.Options()
    options.nsteps = 1000000
    options.atol = 1e-8
    options.rtol = 1e-6

    return psi0, H, x, J, options


def f(psi0, H, x, J, options):
    tlist = np.linspace(0, 10, 11)
    exp_x = qt.mesolve(H, psi0, tlist, J, [x], options=options).expect[0]
    return exp_x


print("Benchmarking:", name)
print("Cutoff: ", end="", flush=True)
checks = {}
results = []
for N in cutoffs:
    print(N, "", end="", flush=True)
    psi0, H, x, J, options = setup(N)
    checks[N] = sum(f(psi0, H, x, J, options))
    t = benchmarkutils.run_benchmark(f, psi0, H, x, J, options, samples=samples, evals=evals)
    results.append({"N": N, "t": t})
print()

benchmarkutils.check(name, checks, 1e-4)
benchmarkutils.save(name, results)

File: benchmarks-QuantumOptics.jl/timeevolution_master_particle.jl
using QuantumOptics
using BenchmarkTools
include("benchmarkutils.jl")

name = "timeevolution_master_particle"

samples = 3
evals = 3
cutoffs = [5:5:60;]

function setup(N)
    xmin = -5
    xmax = 5
    x0 = 0.3
    p0 = -0.2
    sigma0 = 1
    bx = PositionBasis(xmin, xmax, N)
    x = position(bx)
    p = momentum(bx)
    H = p^2 + 2*x^2
    psi0 = gaussianstate(bx, x0, p0, sigma0)
    J = [x + 1im*p]
    psi0, H, x, J
end

function f(psi0, H, x, J)
    T = [0.0:1.0:10;]
    exp_x = Float64[]
    fout(t, rho) = push!(exp_x, real(expect(x, rho)))
    timeevolution.master(T, psi0, H, J; fout=fout, reltol=1e-6, abstol=1e-8)
    exp_x
end

println("Benchmarking: ", name)
print("Cutoff: ")
checks = Dict{Int, Float64}()
results = []
for N in cutoffs
    print(N, " ")
    psi0, H, x, J = setup(N)
    checks[N] = sum(f(psi0, H, x, J))
    t = @belapsed f($psi0, $H, $x, $J) samples=samples evals=evals
    push!(results, Dict("N"=>N, "t"=>t))
end
println()

benchmarkutils.check(name, checks)
benchmarkutils.save(name, results)
